rules/Php74/Rector/Property/TypedPropertyRector.php

```diff
@@ -153,8 +154,7 @@ public function refactor(Node $node): ?Node
             }
         }
 
-        // we are not sure what object type this is
-        if ($varType instanceof NonExistingObjectType) {
+        if ($this->objectTypeAnalyzer->isSpecial($varType)) {
             return null;
         }
```

Comment:
it is not about usage of `UnionType`, it is about usage of `ObjectProphecy` in types found, can be UnionType when has null default value, if no default value, it will be object type of `ObjectProphecy`

Reply:
On php 8.x, without this check, it will produce this for UnionType:

```diff
-    private $obj = null;
+    private ?\Prophecy\Prophecy\ObjectProphecy $obj = null;
```

or object type:

```diff
-    private $obj;
+    private \Prophecy\Prophecy\ObjectProphecy $obj;
```

which incorrect, as it need to be special:

```php
/** @var ReturnString|ObjectProphecy */
```

as original issue described at https://github.com/rectorphp/rector/issues/6843

Reply:
The type should be `ReturnString|ObjectProphecy`. Somewhere we loose the `ReturnString` type.

This is correct result:

```php
private ReturnString|ObjectProphecy|null $obj = null;
```

As a bonus, type might be contextually resolved as never-null, since the `setUp()` method is always run in PHPUnit test.

Reply:
That maybe because of it uses the `$this->prophesize()` call return type so it goes to `ObjectProphecy` type. I will check if it can fallback to the type based on the `@var` if it has `ObjectProphecy` type.

Reply:
I see, that might be difficult to resolve with incorrect 3rd party doc types. Let's go with current solution then :+1: